Themes: Use FluxC ThemeStore in ThemeBrowser

tonyr59h opened the pull request:
[FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/572) must be merged into the feature branch _then_ `develop` _then_ this one will be ready. I suspect there will be changes in FluxC that requires updates on this branch.

Fixes #5122, #6192, and #6459.

These changes update the Theme browser to get Theme data from the new `ThemeStore` in FluxC. Jetpack sites are also supported, hooray!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserActivity.java:
```diff
@@ -184,136 +189,166 @@ public void onSupportSelected(String themeId) {
+        theme.setThemeId(themeId);
+        ThemeStore.ActivateThemePayload payload = new ThemeStore.ActivateThemePayload(mSite, theme);
+
+        if (mSite.isJetpackConnected() && mThemeStore.getInstalledThemeByThemeId(themeId) == null) {
```
Shouldn't we check for `isJetpackConnected` and `isUsingWpComRestApi` at the start of this method? Don't we need that for the `newActivateThemeAction` dispatch?

tonyr59h replied:
We do need the `isUsingWpComRestApi` check at the beginning now that you mention it, `isJetpackConnected` is not required though since .com sites are also supported.

`ThemeAction.ACTIVATE_THEMES` supports both .com and Jetpack sites. The difference with Jetpack sites is that the theme has to be installed first (which is why we have this conditional).

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/WPWebViewActivity.java:
```diff
@@ -366,6 +366,10 @@ public static String getAuthenticationPostData(String authenticationUrl, String
      * @return URL of the login page.
      */
     public static String getSiteLoginUrl(SiteModel site) {
+        if (site.isJetpackConnected()) {
```
This was added in c3ce617d77cb9e9c0c73d301843b6e9a9d4bef59 because I was running into authentication issues with Jetpack sites and `WPWebViewActivity`. So trying to open `Details` or `Support`, for example, would fail for Jetpack sites even though the URL redirects to a public page 🤷‍♂️

kwonye replied:
Reminder that this still hasn't been addressed.

tonyr59h replied:
Moved to `ThemeWebActivity` in 442432a2b125c4959957728c4f62832b56acae4a.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserActivity.java:
```diff
@@ -37,43 +34,49 @@
+    private ThemeModel mCurrentTheme;
     private boolean mIsInSearchMode;
-
+    private boolean mIsFetchingInstalledPlugins;
```
🤦‍♂️  Heh...oops! Must have made that change after reviewing one of your PRs 😆  Fixed in 64ff8f59a0beb81399bbe35b8bb6cdc6649d33c2

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserActivity.java:
```diff
@@ -113,12 +118,6 @@ protected void onPause() {
     }
 
     @Override
-    protected void onRestoreInstanceState(Bundle savedInstanceState) {
```
I don't think there is any difference besides `savedInstanceState` is guaranteed non-null in `onRestoreInstanceState`.

kwonye commented:
@tonyr59h can I get testing steps from you? Want to make sure I don't miss anything.

tonyr59h commented:
> The currently selected theme is no longer shown first

Fixed for WP.com sites in e60886d5ef3b19795b4a852dfe4d6f42c5a93a6e, working on the same for Jetpack. I'm wondering if we should just have the current at the top of the `Uploaded` section, removing the duplicates from the `WordPress.com` section.

> Paid themes are shown.

D'oh! I remember putting a breakpoint in the [existing implementation](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserFragment.java#L324) to see which filter is used (the spinner is always hidden), and it was always `THEME_FILTER_ALL_INDEX` 🤷‍♂️

kwonye commented:
> D'oh! I remember putting a breakpoint in the existing implementation to see which filter is used (the spinner is always hidden), and it was always THEME_FILTER_ALL_INDEX 🤷‍♂️

Hmm, then that's definitely a bug.

Mario seems to have addressed the premium issue in https://github.com/wordpress-mobile/WordPress-Android/pull/5346

So the way I think this should work is:

1. The user has jetpack professional/WP.com business or premium where ALL themes are shown (and the prices hidden for paid themes since it's not going to cost the person anything)
2. The user has jetpack free or premium/WP.com free or personal where only free themes are shown (and Mario's PR lets them buy a theme)

So that filter should still be there imo for us searching.

kwonye commented:
> Try & customize opens chrome. I've noticed that's the behavior in the previous version, but that was definitely changed since V1. Do you know why that is?
> When activating a theme I get the dialog showing "Thanks for choosing null" as text.

These are still showing up, can you address them please? And I'd also like to request again testing steps to make sure I can fully test this feature out for all permutations (jetpack + wp.com accounts)

tonyr59h commented:
> Try & customize opens chrome. I've noticed that's the behavior in the previous version, but that was definitely changed since V1. Do you know why that is?

[Our in-app](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeWebActivity.java#L55) `WebView` doesn't allow local storage access for the Customizer. There was some discussion in #4934 about maybe hiding the `Upload Files` button. We may be able to address that but it's out of scope for this PR.

kwonye commented:
Two more bugs:

1. When returning to the activity, the current theme is not at the top of the list.
2. When clicking details for the current theme the "activate" button shouldn't be at the top right corner.

tonyr59h commented:
> When returning to the activity, the current theme is not at the top of the list.

Fixed via ed9dc9ca20fb82cbeb28d26858b43e15240e9fc9

I also fixed an issue with the way `ThemeStore` updated existing DB models with insufficient data. You'll have to re-install the app on your test device since some of the existing themes may be missing data.

tonyr59h commented:
Testing:

1. Header view current theme
	- [ ] Initially set with correct theme name
	- [ ] Updates when activating new theme

2. Header view buttons
	- [ ] Customize opens device default browser to customize URL for current site
	- [ ] Details opens in-app WebView to details URL for current theme
	- [ ] Support opens in-app WebView to support URL for current theme

3. Activating a theme
	- [ ] Dialog is shown with correct theme name on successful activation
	- [ ] Activated theme is moved to the top of the list

4. Searching
	- [ ] Results are correct (try 'twenty' and there are 8 results)
	- [ ] Empty view is shown when there are no results

5. List
	- [ ] Hides premium themes for non-business/professional sites
	- [ ] Shows active theme first
	- [ ] Jetpack sites have two sections with appropriate themes

kwonye commented:
Running into this when on a jetpack site.

```
E/Volley: [758] BasicNetwork.performRequest: Unexpected response code 403 for https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/mine/?locale=en_US
E/Volley: [756] BasicNetwork.performRequest: Unexpected response code 403 for https://public-api.wordpress.com/rest/v1/sites/73592650/themes/?locale=en_US
E/Volley: [756] BasicNetwork.performRequest: Unexpected response code 403 for https://public-api.wordpress.com/rest/v1/sites/73592650/themes/?locale=en_US
E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1/sites/73592650/themes/?locale=en_US
                 com.android.volley.AuthFailureError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:171)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
E/WordPress-API: Received error response to Jetpack installed themes fetch request.
D/WordPress-API: Dispatching action: ThemeAction-FETCHED_INSTALLED_THEMES
E/WordPress-THEMES: Error fetching themes: This user is not authorized to switch_themes on this blog.
E/Volley: [758] BasicNetwork.performRequest: Unexpected response code 403 for https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/mine/?locale=en_US
E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/mine/?locale=en_US
                 com.android.volley.AuthFailureError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:171)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
E/WordPress-API: Received error response to current theme fetch request.
```

kwonye commented:
Unable to activate a theme on jetpack as well.

- Unable to activate a theme on jetpack site
- Unable to search for themes on jetpack site
- Everytime I go back to the theme activity I get a 'Failed to fetch theme' toast.
- Search doesn't work (results are empty every time no matter what, even with a blank search term which should show all the results)

tonyr59h commented:
> Running into this when on a jetpack site.

We'll have to disable theme's browsing if `switch_themes` isn't allowed for the user, hopefully you've got other accounts to test with ;)

tonyr59h commented:
@kwonye those issues seem to be resolved, I just tested them all.

Also I noticed we weren't cleaning up on logout which could lead to issues if users sign out/in. I added the necessary actions in FluxC and updated this branch in c778de73d8f28f70f017a698f63fcaefac28f2b9.

kwonye commented on WordPress/src/main/java/org/wordpress/android/WordPressDB.java:
```diff
@@ -134,7 +132,6 @@ public WordPressDB(Context ctx) {
                 ctx.deleteDatabase("simperium-store");
                 currentVersion++;
             case 37:
-                ThemeTable.resetTables(db);
```
Shouldn't we delete the tables for users that are upgrading from after version 37?

kwonye commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -682,4 +685,12 @@ public static void addRecentlyPickedSiteId(Integer localId) {
     }
+
+    public static long getLastWpComThemeSync() {
+        return getLong(UndeletablePrefKey.LAST_WP_COM_THEMES_SYNC);
```
Great job with caching!

tonyr59h commented:
> With a jetpack site, I'm unable to do any actions (customize/details/support) on the theme I had activated.

Fixed in 73944da5f47ae869fcf8629e566dfc39a43cf09d. This only happened for Jetpack sites with a theme installed via WP.com which appends `-wpcom` to the end of the theme ID when installed.

> Theme search still doesn't work

This is really weird because my search is working 😕  What device are you using? Do the logs give any info?

kwonye commented:
> This is really weird because my search is working 😕 What device are you using? Do the logs give any info?

Try searching for a theme, then erasing that search term, and then re-searching for another theme.

Here are the logs to help you out:

> E/WordPress-THEMES: Error searching themes: 
> D/WordPress-API: Dispatching action: ThemeAction-SEARCH_THEMES
> D/WordPress-API: Dispatching action: ThemeAction-SEARCH_THEMES
> E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1.2/themes/?search=yo&locale=en_US
>                  com.android.volley.ParseError: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 17966 path $.themes[8].price
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:86)
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125)
>                   Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 17966 path $.themes[8].price
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
>                      at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
>                      at com.google.gson.Gson.fromJson(Gson.java:887)
>                      at com.google.gson.Gson.fromJson(Gson.java:852)
>                      at com.google.gson.Gson.fromJson(Gson.java:801)
>                      at com.google.gson.Gson.fromJson(Gson.java:773)
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:80)
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125) 
>                   Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 17966 path $.themes[8].price
>                      at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) 
>                      at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) 
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82) 
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) 
>                      at com.google.gson.Gson.fromJson(Gson.java:887) 
>                      at com.google.gson.Gson.fromJson(Gson.java:852) 
>                      at com.google.gson.Gson.fromJson(Gson.java:801) 
>                      at com.google.gson.Gson.fromJson(Gson.java:773) 
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:80) 
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125) 
> D/WordPress-API: Receiving OnUnexpectedError event, message: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 17966 path $.themes[8].price
> E/WordPress-API: Received error response to search themes request.
> D/WordPress-API: Dispatching action: ThemeAction-SEARCHED_THEMES
> E/WordPress-THEMES: Error searching themes: 
> E/SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
> E/SpannableStringBuilder: SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
> E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1.2/themes/?search=y&locale=en_US
>                  com.android.volley.ParseError: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 395 path $.themes[0].price
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:86)
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125)
>                   Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 395 path $.themes[0].price
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:224)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
>                      at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82)
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220)
>                      at com.google.gson.Gson.fromJson(Gson.java:887)
>                      at com.google.gson.Gson.fromJson(Gson.java:852)
>                      at com.google.gson.Gson.fromJson(Gson.java:801)
>                      at com.google.gson.Gson.fromJson(Gson.java:773)
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:80)
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125) 
>                   Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 395 path $.themes[0].price
>                      at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:385)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:213)
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) 
>                      at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41) 
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:82) 
>                      at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:129) 
>                      at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:220) 
>                      at com.google.gson.Gson.fromJson(Gson.java:887) 
>                      at com.google.gson.Gson.fromJson(Gson.java:852) 
>                      at com.google.gson.Gson.fromJson(Gson.java:801) 
>                      at com.google.gson.Gson.fromJson(Gson.java:773) 
>                      at org.wordpress.android.fluxc.network.rest.GsonRequest.parseNetworkResponse(GsonRequest.java:80) 
>                      at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:125) 
> D/WordPress-API: Receiving OnUnexpectedError event, message: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 395 path $.themes[0].price
> E/WordPress-API: Received error response to search themes request.

kwonye commented:
Found a few more issues:

1. Going to the end of the themes list results in a progress indicator that doesn't go away. There are obviously no more themes loading but no one would know that. (This doesn't occur in search)
2. Premium themes show up in search results for accounts that can't buy them. Can we filter them out like we do in the normal theme list?
3. Error on activating radcliffe 2 theme with a jetpack site
```
D/WordPress-API: Dispatching action: ThemeAction-ACTIVATE_THEME
E/Volley: [24705] BasicNetwork.performRequest: Unexpected response code 404 for https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/mine/?locale=en_US
E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/mine/?locale=en_US
                 com.android.volley.ServerError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:179)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
D/WordPress-API: Received error response to theme activation request.
D/WordPress-API: Dispatching action: ThemeAction-ACTIVATED_THEME
E/WordPress-THEMES: Error activating theme: The specified theme was not found.
```
4. Jetpack: trying to activate a theme from the WordPress.com themes that's already installed results in an error (with no toast/signal that anything failed)
```
D/WordPress-API: Dispatching action: ThemeAction-INSTALL_THEME
E/Volley: [24704] BasicNetwork.performRequest: Unexpected response code 400 for https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/altofocus-wpcom/install/?locale=en_US
E/WordPress-API: Volley error on https://public-api.wordpress.com/rest/v1.1/sites/73592650/themes/altofocus-wpcom/install/?locale=en_US
                 com.android.volley.ServerError
                     at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:179)
                     at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:114)
D/WordPress-API: Received error response to Jetpack theme installation request.
D/WordPress-API: Dispatching action: ThemeAction-INSTALLED_THEME
E/WordPress-THEMES: Error installing theme: The theme is already installed
```
5. Sites with WP.com Premium don't show premium themes. (Please make sure Jetpack Professional sites show premium themes as well)

tonyr59h commented:
1. Removed progress indicator in a82c412c02861797bad73993e33d8aad99ab29fa
2+5. Fixed via 58f9969e3755f8a72ffc42fa85b4e2e54d4e48e0
4. _Should_ be fixed via ca9a1399bcd14af7cef2f3bbcb14d5298e21f043
3. I'm unable reproduce :/

kwonye commented on WordPress/src/main/java/org/wordpress/android/WordPress.java:
```diff
@@ -126,7 +127,6 @@
     @Inject SiteStore mSiteStore;
     @Inject MediaStore mMediaStore;
 
-
```
This is so minor but thanks so much for doing this.

tonyr59h replied:
lol is always bugs me when I see the double newlines. Happy to oblige with this 😄

kwonye commented on WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserAdapter.java:
```diff
@@ -61,36 +85,90 @@ public ThemeBrowserAdapter(Context context, Cursor c, boolean autoRequery, Theme
+            if (screenshotURL.startsWith("//")) {
+                screenshotURL = screenshotURL.substring(2);
+            }
+            screenshotURL = "https://" + screenshotURL;
```
There is `ThemeWebActivity.THEME_HTTPS_PREFIX` for this.

And can you create one for "http" as well just for cleanliness on line 145 please?

kwonye commented on WordPress/src/main/res/drawable/theme_count_border.xml:
```diff
@@ -0,0 +1,23 @@
+    android:shape="rectangle">
+
+    <stroke
+        android:width="1.5dp"
```
Can this be in `dimens.xml` please?

tonyr59h commented:
All set 👍  Hurry up, Travis!
